File: 1 semester/hw4/4.3/stack.cpp
#include "stack.h"

struct StackElement
{
    TypeElement value;
    StackElement* next;
};

struct Stack
{
    StackElement* head;
};

void push(Stack* stack, TypeElement value)
{
    StackElement* newElement = new StackElement;
    newElement->value = value;
    newElement->next = stack->head;
    stack->head = newElement;
}

TypeElement pop(Stack* stack)
{
    if ((stack->head != nullptr) && !isEmpty(stack)) {
        TypeElement value = stack->head->value;
        StackElement* next = stack->head->next;
        delete stack->head;
        stack->head = next;
        return value;
    }
}

TypeElement top(Stack* stack)
{
    if ((stack->head != nullptr) && !isEmpty(stack))
        return stack->head->value;
}

bool isEmpty(Stack* stack)
{
    return (stack->head == nullptr);
}

Stack* createStack()
{
    Stack *stack = new Stack;
    stack->head = nullptr;
    return stack;
}

void deleteStack(Stack* stack)
{
    while (!isEmpty(stack)) {
        pop(stack);
    }
    delete stack;
}

File: 2 semester/hw1/1.1/src/com/group144/zaicev/Stack.java
package com.group144.zaicev;

public class Stack<Type> {
    private class StackElement {
        public Type value;
        public StackElement next;
        public StackElement(Type value, StackElement next) {
            this.value = value;
            this.next = next;
        }
    }

    private StackElement head = null;

    public void push(Type value) {
        StackElement newElement = new StackElement(value, head);
        head = newElement;
    }

    public boolean isEmpty() {
        return head == null;
    }

    public Type pop() {
        if (isEmpty())
            return null;

        Type value = head.value;
        head = head.next;
        return value;
    }

    public Type top() {
        if (isEmpty())
            return null;
        return head.value;
    }
}
